diff --git a/src/osx/cocoa/glcanvas.mm b/src/osx/cocoa/glcanvas.mm
index 823109cd74..73df24121f 100644
--- a/src/osx/cocoa/glcanvas.mm
+++ b/src/osx/cocoa/glcanvas.mm
@@ -199,8 +199,12 @@ bool wxGLContext::SetCurrent(const wxGLCanvas& win) const
     if ( !m_glContext )
         return false;
 
-    [m_glContext setView: win.GetHandle() ];
-    [m_glContext update];
+    // View can only be set on the UI thread, but context can be set on any thread
+    if (wxIsMainThread())
+    {
+        [m_glContext setView: win.GetHandle() ];
+        [m_glContext update];
+    }
 
     [m_glContext makeCurrentContext];
 
